### PR TITLE
Add auto wrapping support for `DDPFullyShardedStrategy`

### PR DESCRIPTION
**Pull request by rohitgr7**
## What does this PR do?

Follow-up of #14252
adds auto wrapping for fairscale's FSDP.

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: docs/source-pytorch/advanced/model_parallel.rst [hunk 1]
```diff
@@ -114,9 +118,27 @@ To reach larger parameter sizes and to be memory efficient, we have to shard par
     This is a limitation of Fully Sharded Training that will be resolved in the future.
 
 Enabling Module Sharding for Maximum Memory Efficiency
-""""""""""""""""""""""""""""""""""""""""""""""""""""""
+------------------------------------------------------
+
+Auto Wrapping
```
Comment (rohitgr7) [inline docs/source-pytorch/advanced/model_parallel.rst:123]: this is the only notable doc update. Rest is just header type update for better sidebar.

File: src/pytorch_lightning/strategies/fully_sharded.py [hunk 2]
```diff
@@ -26,16 +27,25 @@
...
     from fairscale.nn import default_auto_wrap_policy, enable_wrap
     from fairscale.nn.data_parallel import FullyShardedDataParallel
+else:
+    FullyShardedDataParallel = None
 
 log = logging.getLogger(__name__)
 
 
+class _DDPFullyShardedStrategyModuleWrapper(_LightningModuleWrapperBase):
+    def state_dict(self, *args: Any, **kwargs: Any) -> Dict[str, Any]:  # type: ignore[override]
```
Comment (rohitgr7) [inline src/pytorch_lightning/strategies/fully_sharded.py:45]: this is required because with FSDP `self.lightning_module.state_dict()` is empty. So we need to call `self.trainer.model.state_dict` (wrapped version) and use this wraper to avoid extra keys `_forward_module.layer...` since we want `layer...` in state_dict.

ref: https://fairscale.readthedocs.io/en/stable/_modules/fairscale/nn/data_parallel/fully_sharded_data_parallel.html#FullyShardedDataParallel.state_dict

File: src/pytorch_lightning/strategies/fully_sharded.py [hunk 3]
```diff
@@ -132,6 +142,25 @@ def process_group(self) -> Any:
...
+        """Returns model state."""
+        assert self.model is not None
+        return self.model.state_dict()
+
+    def connect(self, model: "pl.LightningModule") -> None:
+        """Called by the accelerator to connect the accelerator and the model with this plugin."""
+        # TODO: Wait for this issue to resolve and remove this blocker
+        # https://github.com/facebookresearch/fairscale/issues/648
+        # Also make sure to update the tests
+        if not is_overridden("configure_sharded_model", self.lightning_module) and len(list(model.parameters())) == 0:
```
Comment (rohitgr7) [inline src/pytorch_lightning/strategies/fully_sharded.py:155]: because of https://github.com/facebookresearch/fairscale/issues/648
this is a blocker, but the user can still reinitialize the model and CPU, which will work. Added tests for both cases.

File: src/pytorch_lightning/strategies/fully_sharded.py [hunk 4]
```diff
@@ -144,17 +173,46 @@ def setup(self, trainer: "pl.Trainer") -> None:
...
 
-        self.setup_precision_plugin()
         self.configure_ddp()
+        assert isinstance(self.model, pl.LightningModule)
+        self.model = _DDPFullyShardedStrategyModuleWrapper(self.model)
+        assert self.lightning_module is not None
+        if not is_overridden("configure_sharded_model", self.lightning_module):
+            self.model = self._setup_model(self.model)
+        self.setup_optimizers(self.lightning_module.trainer)
+        optimizers_to_device(self.optimizers, self.root_device)
```
Comment (rohitgr7) [inline src/pytorch_lightning/strategies/fully_sharded.py:187]: flow:
```
- move model to device is necessary
- wrap it
- initialize optimizers
```
this is recommended by fsdp and it's examples

File: src/pytorch_lightning/strategies/fully_sharded.py [hunk 5]
```diff
@@ -144,17 +173,46 @@ def setup(self, trainer: "pl.Trainer") -> None:
...
+            " you will get an error.\ntorch.optim.Optimizer(self.trainer.model.parameters(), ...)"
+        )
+
+        return FullyShardedDataParallel(
+            module=model,
+            process_group=self.process_group,
+            cpu_offload=self.cpu_offload,
+            move_grads_to_cpu=self.move_grads_to_cpu,
+            flatten_parameters=self.flatten_parameters,
+            mixed_precision=(self.precision_plugin.precision in (PrecisionType.MIXED, PrecisionType.HALF)),
```
Comment (rohitgr7) [inline src/pytorch_lightning/strategies/fully_sharded.py:208]: mixed precision and full precision are different in fsdp. I will update in follow-up.

File: src/pytorch_lightning/strategies/fully_sharded.py [hunk 6]
```diff
@@ -190,35 +248,29 @@ def configure_ddp(self) -> None:
             # (TODO: need to figure out solution)
             self.model_to_device()
```
Comment (rohitgr7) [inline src/pytorch_lightning/strategies/fully_sharded.py:249]: this is incorrect too. We need to move the model to the device only in case of `ssd-offload`. I will do it in a follow-up.

File: docs/source-pytorch/advanced/model_parallel.rst [hunk 7]
```diff
@@ -114,9 +118,37 @@ To reach larger parameter sizes and to be memory efficient, we have to shard par
...
+^^^^^^^^^^^^^
+
+Model layers should be wrapped in FSDP in a nested way to save peak memory and enable communication and computation overlapping. The
+simplest way to do it is auto wrapping, which can serve as a drop-in replacement for DDP without changing the rest of the code. You don't
+have to ``wrap`` layers manually as in the case of manual wrapping.
+
+.. note::
+    While initializing the optimizers inside ``configure_optimizers`` hook, make sure to use ``self.trainer.model.parameters()``, else
+    PyTorch will raise an error. This is required because when you use auto-wrap, the model layers are sharded and your
+    ``lightning_module.parameters()`` will return a generator with no params.
```
Comment (awaelchli) [inline docs/source-pytorch/advanced/model_parallel.rst:133]: Oh no, this is a bummer. I think we should find a way to work around this. Maybe write: "This inconvenience will be addressed in the future".

**Review by awaelchli — APPROVED** [review]
great work discovering and handling these edge cases!

**Review by carmocca — APPROVED** [review]
Great job! Looking forward to those followups
